pkg/controller/metrics/metrics.go

```diff
@@ -18,14 +18,90 @@ import (
 	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1"
 	hiveintv1alpha1 "github.com/openshift/hive/pkg/apis/hiveinternal/v1alpha1"
 	"github.com/openshift/hive/pkg/constants"
+	controllerutils "github.com/openshift/hive/pkg/controller/utils"
 	"github.com/openshift/hive/pkg/imageset"
 )
 
 const (
 	ControllerName = hivev1.MetricsControllerName
 )
 
+// metrics collected through custom prometheus collector
+type customCollector struct {
+	client client.Client
+
+	// metricClusterDeploymentProvisionUnderwaySeconds is a prometheus metric for the number of seconds
+	// between when a still provisioning cluster was created and now.
+	metricClusterDeploymentProvisionUnderwaySeconds *prometheus.Desc
+}
+
+// collects the metrics for custom collector
+func (cc customCollector) Collect(ch chan<- prometheus.Metric) {
+	ccLog := log.WithField("controller", "metrics")
+	ccLog.Info("calculating metrics across all ClusterDeployments")
+	// Load all ClusterDeployments so we can accumulate facts about them.
+	clusterDeployments := &hivev1.ClusterDeploymentList{}
+	err := cc.client.List(context.Background(), clusterDeployments)
+	if err != nil {
+		log.WithError(err).Error("error listing cluster deployments")
+	} else {
+		accumulator, err := newClusterAccumulator(infinity, []string{"0h", "1h", "2h", "8h", "24h", "72h"})
+		if err != nil {
+			ccLog.WithError(err).Error("unable to calculate metrics")
+			return
+		}
+		for _, cd := range clusterDeployments.Items {
```

Comment:
Your implementation looks a little different than in https://stackoverflow.com/a/57308144 where they maintain the data in an external map. Initially I was a little worried that this Collect function is called every time we're scraped, but I'm not sure that's the case, I actually can't tell how often this is called. I'm a little worried about iterating every ClusterDeployment, but otoh they should all be cached in memory anyhow, so I'm not sure how much of an issue it is even if this is called every 15 seconds. (which it doesn't seem to be)

Your implementation has a nice benefit where we don't need something external updating a shared map with the metrics data, locks etc. Metrics disappear as soon as the Collect method runs so delete is no longer an issue.

Can you track down when Collect will be called so we can understand it and make a final call? As it stands I like this and I hope we can stick with it as it is.